src/tree_types.ml

```diff
@@ -0,0 +1,265 @@
+(** Miscellaneous types and utilities for decision trees. *)
+open Extra
+
+(** {3 Tree constructors} *)
+
+(** Define a simpler representation of terms. *)
+
+(** A constructor is the representation of a symbol along with the number of
+    arguments to which it is applied. *)
+type treecons =
+  | Symb of
+      { c_mod : Files.module_path
+      (** Module name where the symbol of the constructor is defined. *)
+      ; c_sym : string
+      (** Symbol of the constructor. *)
+      ; c_ari : int
+      (** Arity of the considered constructor.  A same symbol representation
+          may generate several constructors with different arities.  For
+          instance, with [g: N ⇒ N ⇒ N], [rule f g (g &x) → r] uses [g] with
+          arity 1 and 2. *) }
+  | Abst
+  (** An abstraction*)
+  | Vari of string
+  (** A bound variable with a name. *)
+
+(** [tc_pp o c] prints tree constructor [c] to output channel [o]. *)
+let tc_pp : treecons pp = fun oc -> function
+  | Abst    -> Format.fprintf oc "λ"
+  | Vari(s) -> Format.pp_print_string oc s
+  | Symb(t) -> Format.fprintf oc "%s<:%d" t.c_sym t.c_ari
+
+(** [tc_compare c d] is a comparison function for constructors; more efficient
+    than the pervasive. *)
+let tc_compare : treecons -> treecons -> int = fun ca cb ->
+  match ca, cb with
+  | Symb(a), Symb(b) ->
+      begin match Int.compare a.c_ari b.c_ari with
+      | 0 ->
+          begin match String.compare a.c_sym b.c_sym with
+          | 0 -> Pervasives.compare a.c_mod b.c_mod
+          | x -> x
+          end
+      | x -> x
+      end
+  | x, y                 -> Pervasives.compare x y
+
+(** A simple comparable type. *)
```

Comment:
This is useless, just define the threshold in a global variable (we could even make it configurable with a command-line flag), and then used the `Treecons` module (with module type `Map.OrderedType`) in the functor.